src/compile/data/bin.ts

```diff
@@ -1,18 +1,32 @@
 import {BinParams, binToString} from '../../bin';
-import {Channel} from '../../channel';
+import {Channel, NONSPATIAL_SCALE_CHANNELS, SPATIAL_SCALE_CHANNELS} from '../../channel';
 import {Config} from '../../config';
 import {field, FieldDef, normalizeBin} from '../../fielddef';
 import {BinTransform} from '../../transform';
-import {Dict, duplicate, extend, flatten, keys, vals} from '../../util';
+import {NOMINAL, ORDINAL} from '../../type';
+import {contains, Dict, duplicate, extend, flatten, keys, vals} from '../../util';
 import {VgBinTransform, VgTransform} from '../../vega.schema';
 import {binFormatExpression} from '../common';
 import {isUnitModel, Model, ModelWithField} from '../model';
 import {DataFlowNode} from './dataflow';
 
-function rangeFormula(model: ModelWithField, fieldDef: FieldDef<string>, channel: Channel, config: Config) {
-    const discreteDomain = model.hasDiscreteDomain(channel);
+/**
+ * Checks whether a fieldDef for a particular channel requires a computed bin range.
+ */
+export function binRequiresRange(fieldDef: FieldDef<string>, channel: Channel) {
+  if (!fieldDef.bin) {
```

Comment:
Wonder if this is necessary -- we could also just take type as input.

Reply:
I liked this better as fieldDef is more standard and this way I can add a sanity check.